GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+* With that in mind it becomes worth to **invest more in integration tests** instead of unit tests
+  * which **reduces the use of mocks** and
+  * leads to **more realistic tests**
+
+## One-by-one steps
+
+* Create a module testing configuration in the test sources for each module in your project (e.g. for a module `app` create a file `AppModuleTestingConfiguration` with the following code: `val appModuleTestingConfiguration = moduleTestingConfiguration { ... }`; see below for more info on how to add dependencies to the configuration)
+  * Always use the same package structure for test sources, e.g. if you app resides in a module `app` and has a base package `org.myorg.product.app` then put the module test configuration in exactly in the same package
+  * If there is a class `org.myorg.product.app.view.LoginViewModel` then the test class `LoginViewModelTest` and `LoginViewModelSteps` would reside in the exact same package, too
+  * If you have a dedicated module for test sources you should still stick to the same package structure, e.g. if the module is `:app:test` _don't_ add `test` to the package structure for test sources
+  * If modules depend on each other the module testing configuration should do the same, e.g. `val appModuleTestingConfiguration = moduleTestingConfiguration(dependentConfig1, dependentConfig2, ...) { ... }`, this automatically imports all dependent configurations
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
```

Comment:
On the pump we tried to name test functions after requirements. So we don't name _scenarios_, but express what is expected to happen. And I think that would make sense here as well that would be e.g.
instead of `New user trying to log in` → `New user is asked for name, nationality and password`.
instead of `Existing user trying to log in` → `Existing user is only asked for password`.
instead of `Weather condition is cloudy` → `Background turns gray when weather condition is cloudy`.
What do you think?

Reply:
i like the proposals, but would put it as "if it makes sense", especially to distinguish test function from each other. i want to add them as better examples, thanks!
i'll make a differentiated proposal also depending on the outcome of [that discussion](https://github.com/mysugr/sweetest/pull/24#discussion_r401611362). could you please add your opinion there?

Reply:
I would put the naming conventions in an own section and then structure it similar to our [Variable guidelines](https://github.com/mysugr/shared-android/blob/develop/documentation/naming-conventions/variables.md). Otherwise it is a bit all over the place in the document. So you can add wrong, ok and good examples. I agree with Daniel that test names should at least define what is expected to happen.